File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  const findMatchIndex = element => arr.indexOf(element)

  const matchTest = mappedArray => mappedArray.filter(indexValue => indexValue !== -1)

  const firstMatch = filteredArray => filteredArray.find((element, index) => element == filteredArray[index + 1])

  const matchFind = filteredArray => filteredArray.length > 1 ? arr[firstMatch(filteredArray)] : -1

  const arrayMatchFinder = arr => matchFind(matchTest(arr.map(element => findMatchIndex(element))))

  return arrayMatchFinder(arr)    
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
/*
  -map the indexes of elements within the array that match, if no match is given the value for that elements index in the map will be -1
  -filter out this map so that it contains only elements that are NOT -1
  -if this filtered map has a length of 0 (no elements that are not of index -1) return -1, else return the value of the original
   array at the index of the first value within the filtered map.
*/
// And a written explanation of your solution
/*
  In order to overcome the synchronus nature of javascript I had to create multiple functions that broke the solution down into steps.
    each one called another one with the desired result being given at the end.  First I created an array of the indexes of all the
    matches within an array.  If there were none I would return -1 but if their were multiple matches I had to call yet another
    function that found the first index that repeated one after the other.  I would then use this index value as the return index
    of the initial array.
*/

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # type your code in here
  matches = arr.map.with_index do |value, index|
    if value == arr[index + 1]
      index
    end
  end
  match_index = matches.select {|element| !!element}
  (match_index.length > 0 ? arr[match_index[0]] : -1)

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
=begin
  -create a map of the array containing the indexes of all the matching elements that occur at the next index
  -create another array that contains all the elements of this map that passed initial test.
  -make sure this new array has a length greater than 0, return the original array at the first index of this new array
=end
# And a written explanation of your solution
=begin
  I had to create an array of indexes in which a duplicate element is found in the original array that corresponds with the following
  index of the original array.  I then had to select out of this list to exclude false values (elements that had no duplicate or
  did not dupicate in sequence).  I then had to return the value of the original array at this arrays first index.  If there wasnt
  one, I needed to return -1. 
=end
